Fix 3D character snap on moving platforms

pouleyKetchoupp opened the pull request:
Fixes the first point from https://github.com/godotengine/godot/issues/50732#issuecomment-896292908 (regression from #51457).
CC @TokageItLab
CC @fabriceci

Applying the platform velocity when leaving the platform floor should be done after snapping to keep things consistent, like it's done in 2D.

This issue also affects the 3.x version, so I'll make a separate PR for 3.x.

akien-mga reviewed (approved):
Needs rebase, then can be merged.

pouleyKetchoupp committed:
Fix 3D character snap on moving platforms

Applying the platform velocity when leaving the platform floor should be
done after snapping to keep things consistent, like it's done in 2D.
